sycl/doc/extensions/proposed/sycl_ext_oneapi_prefetch.asciidoc

```diff
@@ -0,0 +1,454 @@
+(e.g. a request to prefetch into L1 and L4 is treated as a request to prefetch
+into L1). When no cache levels are specified, the default behavior is to
+prefetch into the lowest level cache (i.e. L1).
+
+[NOTE]
+====
+Future hints may alter the default behavior of prefetches with respect to cache
+levels. Any such alterations are expected to be documented in the definition
+of those new hints.
+====
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+enum class cache_level {
+  L1,
+  L2,
+  L3,
+  L4,
+};
+
+struct nontemporal;
+
+struct prefetch_hint_key {
+  template <cache_level Level, typename Hint>
+  using value_t = property_value<prefetch_hint_key,
+                                 std::integral_constant<cache_level, Level>,
+                                 Hint>
+};
+
+template <cache_level Level, typename Hint>
+inline constexpr prefetch_hint_key::value_t<Level, Hint> prefetch_hint;
+
+inline constexpr prefetch_hint_key::value_t<cache_level::L1, void> prefetch_hint_L1;
+inline constexpr prefetch_hint_key::value_t<cache_level::L2, void> prefetch_hint_L2;
+inline constexpr prefetch_hint_key::value_t<cache_level::L3, void> prefetch_hint_L3;
+inline constexpr prefetch_hint_key::value_t<cache_level::L4, void> prefetch_hint_L4;
+
+inline constexpr prefetch_hint_key::value_t<cache_level::L1, nontemporal> prefetch_hint_L1_NT;
+inline constexpr prefetch_hint_key::value_t<cache_level::L2, nontemporal> prefetch_hint_L2_NT;
+inline constexpr prefetch_hint_key::value_t<cache_level::L3, nontemporal> prefetch_hint_L3_NT;
+inline constexpr prefetch_hint_key::value_t<cache_level::L4, nontemporal> prefetch_hint_L4_NT;
+
+} // namespace sycl::ext::oneapi::experimental
+----
+
+If the `nontemporal` type is used as the value of the `Hint` template
+parameter, this acts as a hint that the data being prefetched will not be
+reused.
+
+[NOTE]
+====
+The implementation of `nontemporal` is device-specific, owing to differences in
+cache hierarchies and replacement policies. On some devices non-temporal data
+may be evicted earlier than temporal data, while on other devices non-temporal
+data may not be stored in certain levels of cache at all.
+====
+
+
+=== Work-item prefetches
+
+The functions in this section allow individual work-items to prefetch data.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename Properties = ext::oneapi::experimental::properties<>>
+void prefetch(void* ptr, Properties properties = {});
+
+template <typename Properties = ext::oneapi::experimental::properties<>>
+void prefetch(void* ptr, size_t bytes, Properties properties = {});
+
+template <typename T, typename Properties = ext::oneapi::experimental::properties<>>
+void prefetch(T* ptr, Properties properties = {});
+
+template <typename T, typename Properties = ext::oneapi::experimental::properties<>>
+void prefetch(T* ptr, size_t count, Properties properties = {});
```

Comment:
What is our policy on raw pointers vs. `muilti_ptr`?  Should all these functions take `multi_ptr` instead?

Reply:
I think we want to take both.  The group algorithms accept either raw pointers or `multi_ptr`.

I'm not sure how to do this cleanly, though.  Now that you've asked the question, I realized that we should also limit this to _global_ pointers.  Do you have a preference for either of the below (or any other suggestions)?

Some constraints magic:
```
template <typename Ptr, typename Properties = ext::oneapi::experimental::properties<>>
void prefetch(Ptr ptr, size_t count, Properties properties = {});

_Constraints_: `Ptr` is `void*` or a `multi_ptr<void>`.
_Preconditions_: `ptr` points to an object in the global address space.
```

Another overload and fewer constraints:
```
template <typename Properties = ext::oneapi::experimental::properties<>>
void prefetch(void* ptr, size_t count, Properties properties = {});

_Preconditions_: `ptr` points to an object in the global address space.

...

template <access::address_space AddressSpace, access::decorated Decoration, typename Properties = ext::oneapi::experimental::properties<>>
void prefetch(multi_ptr<void, AddressSpace, Decoration> ptr, size_t count, Properties properties = {});

_Constraints_: `AddressSpace` is `global_space` or `generic_space`.
_Preconditions_: `ptr` points to an object in the global address space.
```

I'm leaning towards the extra overloads.  Even if they're painful to specify, they seem easier to understand.

Reply:
Yes, the overloads seem better to me too.

Reply:
I've added all the `multi_ptr` overloads in 9929ebf9597d89bba192cf8a929ead5e27e6cf6d.  There was a lot of copying and pasting involved, so please double-check that I didn't make any mistakes here.